src/builder/create_attachment.rs

```diff
@@ -117,4 +106,174 @@ impl CreateAttachment {
+/// msg.edit(ctx, EditMessage::new().attachments(
+///     EditAttachments::new_keep_all(&msg).add(my_attachment)
+/// )).await?;
+/// # Ok(()) }
+/// ```
+///
+/// ## Delete all but the first attachment
+///
+/// ```rust,no_run
+/// # use serenity::all::*;
+/// # async fn _foo(ctx: Context, mut msg: Message, my_attachment: CreateAttachment) -> Result<(), Error> {
+/// msg.edit(ctx, EditMessage::new().attachments(
+///     EditAttachments::new().keep(msg.attachments[0].id)
+/// )).await?;
+/// # Ok(()) }
+/// ```
+///
+/// ## Delete only the first attachment
+///
+/// ```rust,no_run
+/// # use serenity::all::*;
+/// # async fn _foo(ctx: Context, mut msg: Message, my_attachment: CreateAttachment) -> Result<(), Error> {
+/// msg.edit(ctx, EditMessage::new().attachments(
+///     EditAttachments::new_keep_all(&msg).dont_keep(msg.attachments[0].id)
+/// )).await?;
+/// # Ok(()) }
+/// ```
+///
+/// # Notes
+///
+/// Internally, this type is used not just for message editing endpoints, but also for message
+/// creation endpoints.
+#[derive(Default, Debug, Clone, serde::Serialize)]
+#[serde(transparent)]
+#[must_use]
+pub struct EditAttachments {
+    new_and_existing_attachments: Vec<NewOrExisting>,
+}
+
+impl EditAttachments {
+    /// An empty attachments builder.
+    ///
+    /// Existing attachments are not kept by default, either. See [`Self::new_keep_all()`] or
+    /// [`Self::keep()`].
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Creates a new attachments builder that keeps all existing attachments.
+    ///
+    /// Shorthand for [`Self::new()`] and calling [`Self::keep()`] for every
+    /// [`AttachmentId`] in [`Message::attachments`].
+    ///
+    /// If you only want to keep a subset of attachments from the message, either implement this
+    /// method manually, or use [`Self::dont_keep()`].
+    ///
+    /// **Note: this EditAttachments must be run on the same message as is supplied here, or else
+    /// Discord will throw an error!**
+    pub fn new_keep_all(msg: &Message) -> Self {
+        Self {
+            new_and_existing_attachments: msg
+                .attachments
+                .iter()
+                .map(|a| {
+                    NewOrExisting::Existing(ExistingAttachment {
+                        id: a.id,
+                    })
+                })
+                .collect(),
+        }
+    }
+
+    /// This method adds an existing attachment to the list of attachments that are kept after
+    /// editing.
+    pub fn keep(mut self, id: AttachmentId) -> Self {
+        self.new_and_existing_attachments.push(NewOrExisting::Existing(ExistingAttachment {
+            id,
+        }));
+        self
+    }
+
+    /// This method removes an existing attachment from the list of attachments that are kept after
+    /// editing.
+    pub fn dont_keep(mut self, id: AttachmentId) -> Self {
```

Comment:
I think a better name would be `remove`, or maybe `drop`.

Reply:
Your envisioned mental model of EditAttachments seems to be: two seperate lists of attachments, initially empty, that each can be pushed to. This more closely matches internals than my envisioned mental model of EditAttachments, which is: one list of new attachments that is initially empty and can be pushed to; and a set of all existing attachments, initially full but with all items set "inactive", where you don't push to the set but rather activate specific attachments by their ID (think `existing_attachments.set_to_keep(id, true);` rather than `existing_attachments.add_to_existing(id);`)

My idea is that the latter mental model is more intuitive to people who don't happen to already know how the API works under the hood. And that's also why I chose the terms "keep"/"dont_keep"

Reply:
I think the name `remove` still fits with this mental model, as it marks the existing attachment to be explicitly "removed" or otherwise "excluded" from the message. And I'd say that I do in fact view the existing attachments as a set and not as an empty list. I just really dislike the use of `dont` as part of the method like this. It's normally a contraction but obviously you can't put an apostrophe in a method name (I'm not sure if that would make it a worse name or better tbh lol). As it stands it reads not seriously at all.

In principle, the pair of `keep`/`dont_keep` sounds too passive for me. If it were `keep`/`remove`, that would be better. The pair `keep`/`drop` has a nice symmetry, but `drop` is not a very conventional term, so I like it less. There may be some replacement for `keep` that's also less passive, but I can't think of one at the moment.